### PR TITLE
store associated production schema when creating non-production schema

### PR DESCRIPTION
**Pull request by JoanneBogart**
Linkage between a non-production schema and a production schema is determined when the non-production schema is created.  Save the production schema name in the Provenance table of the non-production schema during creation. Fetch it when a connection to a schema is made.

### CODE REVIEW

File: src/dataregistry/schema/schema.yaml [hunk 1]
```diff
@@ -216,10 +220,13 @@ dataset:
...
     description: "UID of person who registered the entry"
     nullable: False
-  access_API:
+  access_api:
     type: "StringShort"
     description: "Describes the software that can read the dataset (e.g., 'gcr-catalogs', 'skyCatalogs')"
     cli_optional: True
+  access_api_configuration:
+    type: "String"
+    description: "Additional (text) info which may be needed by access_API"
```
Comment (stuartmcalpine) [inline src/dataregistry/schema/schema.yaml:229]: Lowercase the API in the description also

File: src/dataregistry/db_basic.py [hunk 2]
```diff
@@ -173,13 +169,25 @@ def __init__(self, db_connection, get_db_version=True):
...
+        # Fetch and save db versioning, assoc. production schema
+        # if present and requested
+        self._prod_schema = None
         if db_connection.dialect == "sqlite":
             prov_name = "provenance"
         else:
             prov_name = ".".join([self._schema, "provenance"])
+
+            # prov_table = self._metadata.tables[prov_name]
         if prov_name in self._metadata.tables and get_db_version:
```
Comment (stuartmcalpine) [inline src/dataregistry/db_basic.py:181]: More general comment, is there any reason why `prov_name` shouldn't in the list of tables, i.e.,

`if prov_name in self._metadata.tables`

If this is not true, shouldnt an error be raised?

File: src/dataregistry/db_basic.py [hunk 3]
```diff
@@ -173,13 +169,25 @@ def __init__(self, db_connection, get_db_version=True):
...
-        # Fetch and save db versioning if present and requested
+        # Fetch and save db versioning, assoc. production schema
+        # if present and requested
+        self._prod_schema = None
         if db_connection.dialect == "sqlite":
             prov_name = "provenance"
         else:
             prov_name = ".".join([self._schema, "provenance"])
+
+            # prov_table = self._metadata.tables[prov_name]
```
Comment (stuartmcalpine) [inline src/dataregistry/db_basic.py:180]: Left over comment? Can delete?

File: src/dataregistry/db_basic.py [hunk 4]
```diff
@@ -173,13 +169,25 @@ def __init__(self, db_connection, get_db_version=True):
...
         if prov_name in self._metadata.tables and get_db_version:
             prov_table = self._metadata.tables[prov_name]
+            stmt = select(column("associated_production")).select_from(prov_table)
+            stmt = stmt.order_by(prov_table.c.provenance_id.desc())
+            with self._engine.connect() as conn:
+                results = conn.execute(stmt)
+                conn.commit()
```
Comment (stuartmcalpine) [inline src/dataregistry/db_basic.py:187]: Don't need `commit` for queries (I think)

Can also remove the commit below if true

File: src/dataregistry/db_basic.py [hunk 5]
```diff
@@ -242,12 +250,17 @@ def _insert_provenance(
         One of "create", "migrate"
     comment : str, optional
         Briefly describe reason for new version
+    associated_production : str, defaults to "production"
+        Namae of production schema, if any, this schema may reference
```
Comment (stuartmcalpine) [inline src/dataregistry/db_basic.py:254]: "Name"

**Review by stuartmcalpine — CHANGES_REQUESTED** [review]
Looks fine overall, left some comments in the review.

In addition:

- Remove GitPython from dependencies from pyproject.toml?
- CI test for new column in prov?
- Update changelog and code version
- Update database version

File: src/dataregistry/db_basic.py (hunk 2 above)
Comment (JoanneBogart) [reply src/dataregistry/db_basic.py:181]: I guess so.  This is not code I changed. Perhaps the thinking was then that there should be support for old-style databases with no provenance table, but there is no reason for it now.

File: src/dataregistry/db_basic.py (hunk 3 above)
Comment (JoanneBogart) [reply src/dataregistry/db_basic.py:180]: Done

**Comment by JoanneBogart** [comment]
I think I've addressed everything except for adding a unit test for the new db field.  I'm not sure such a test would accomplish much.  In fact I'd like to get rid of test_database.test_db_version (the only way it can fail is if you forget - as I did - to update the test.  It doesn't really test the code).  I'd also like to get rid of test_database_functions.test_db_version.  Here there is something to test - whether the database version of the current code is compatible with the database you're connecting to - but the test as it stands now, since the code version running the query is known to be the same as the one that created the db - doesn't really check that.  

Instead there should be a check when people use the code for compatibility: major versions of db should match and minor version db for the code running should be >= minor version of the db it's connecting to.  This will take some rearranging to make the db versions associated with the running code, currently buried in create_registry_schema.py, accessible to TableMetadata.

I can do that in this PR or a separate one.

**Comment by stuartmcalpine** [comment]
>  I'd also like to get rid of test_database_functions.test_db_version.

I'm happy for that "test" to go, it doesn't do anything as you say

File: src/dataregistry/db_basic.py (hunk 4 above)
Comment (JoanneBogart) [reply src/dataregistry/db_basic.py]: Yes.  Now removed.

**Review by stuartmcalpine — APPROVED** [review]

**Pull request merged**